Adding lazy loading support for dataset ID.

dhermes opened the pull request:
**NOTE**: Has #666 as diffbase.

Still need to support connection (and eventually do the same
in storage).

Again verified only the lazy loading tests used implicit
behavior via

```diff
diff --git a/gcloud/datastore/_implicit_environ.py b/gcloud/datastore/_implicit_environ.py
index 6e61636..6d67be6 100644
--- a/gcloud/datastore/_implicit_environ.py
+++ b/gcloud/datastore/_implicit_environ.py
@@ -177,6 +177,10 @@ class _LazyProperty(object):
         self._method = method

     def __get__(self, obj, objtype):
+        class FooError(Exception):
+            pass
+        raise FooError
+
         if obj is None or objtype is not _DefaultsContainer:
             return self
```

dhermes commented:
@tseaver I encountered your favorite, `pylint` made me change my code.

I first made a decorator

```python
def _lazy_property_deco(method):
    return _LazyProperty(method.__name__, method)

class _DefaultsContainer(object):
    @_lazy_property_deco
    def dataset_id():
        return _determine_default_dataset_id()
```

but `pylint` complained of [method-hidden (E0202)][1]:

> Used when a class defines a method which is hidden by an instance
> attribute from an ancestor class or set by some client code.

Essentially the decorator might put something different in the class definition than the code indicates.

If you prefer this approach we can add `method-hidden` to `[MESSAGES CONTROL].disable` in [`pylintrc_default`][2].

[1]: http://docs.pylint.org/features.html#id20
[2]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/72b6359c51aafab7d8beed2521fb468b8e442bca/pylintrc_default

tseaver commented:
Shadowing a method with an instance attribute is a perfectly normal, acceptable practice:  claiming it is wrong is insisting we ride the Python bike with training wheels on.  Let's disable the warning.

dhermes commented:
@tseaver PTAL. I just rebased this on top of #666.

Some things which differ than the original PR:

- I added a commit to fix the broken build (was swapping out the env var name in the wrong module)
- I added a commit to disable the warning and switch back to using a method decorator and also addressed your [comment][1] in #668 

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/pull/668/files#r25011223

tseaver commented:
LGTM.
